core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java

```diff
@@ -138,76 +155,106 @@ void setDelegate(LoadBalancer lb) {
-      // Check for balancer addresses
-      boolean haveBalancerAddress = false;
-      for (EquivalentAddressGroup s : servers) {
-        if (s.getAttributes().get(GrpcAttributes.ATTR_LB_ADDR_AUTHORITY) != null) {
-          haveBalancerAddress = true;
-          break;
-        }
+  /**
+   * Picks a load balancer based on given criteria.  In order of preference:
+   *
+   * <ol>
+   *   <li>User provided lb on the channel.  This is a degenerate case and not handled here.</li>
+   *   <li>"grpclb" if any gRPC LB balancer addresses are present</li>
+   *   <li>The policy picked by the service config</li>
+   *   <li>"pick_first" if the service config choice does not specify</li>
+   * </ol>
+   *
+   * @param servers The list of servers reported
+   * @param config the service config object
+   * @return the new load balancer factory, never null
+   */
+  @VisibleForTesting
+  static PolicySelection decideLoadBalancerProvider(
+      List<EquivalentAddressGroup> servers, @Nullable Map<String, Object> config,
+      ChannelLogger channelLogger)
+    throws PolicyException {
+    // Check for balancer addresses
+    boolean haveBalancerAddress = false;
+    for (EquivalentAddressGroup s : servers) {
+      if (s.getAttributes().get(GrpcAttributes.ATTR_LB_ADDR_AUTHORITY) != null) {
+        haveBalancerAddress = true;
+        break;
       }
+    }
 
-      if (haveBalancerAddress) {
-        return getProviderOrThrow("grpclb", "NameResolver has returned balancer addresses");
-      }
+    if (haveBalancerAddress) {
+      LoadBalancerProvider provider =
+          getProviderOrThrow("grpclb", "NameResolver has returned balancer addresses");
+      return new PolicySelection(provider, null);
+    }
 
-      String serviceConfigChoiceBalancingPolicy = null;
-      if (config != null) {
-        serviceConfigChoiceBalancingPolicy =
-            ServiceConfigUtil.getLoadBalancingPolicyFromServiceConfig(config);
-        if (serviceConfigChoiceBalancingPolicy != null) {
-          // Handle ASCII specifically rather than relying on the implicit default locale of the str
-          return getProviderOrThrow(
-              Ascii.toLowerCase(serviceConfigChoiceBalancingPolicy),
-              "service-config specifies load-balancing policy");
+    if (config != null) {
+      List<Map<String, Object>> lbConfigs =
+          ServiceConfigUtil.getLoadBalancingConfigsFromServiceConfig(config);
+      LinkedHashSet<String> policiesTried = new LinkedHashSet<String>();
+      for (Map<String, Object> lbConfig : lbConfigs) {
+        if (lbConfig.size() != 1) {
+          throw new PolicyException(
+              "There are " + lbConfig.size()
+              + " load-balancing configs in a list item. Exactly one is expected. Config="
+              + lbConfig);
         }
+        Entry<String, Object> entry = lbConfig.entrySet().iterator().next();
+        // Internally the service config is written in Protobuf and converted to JSON.
```

Comment:
Can this be done in the internal NameResolver instead?

I'd prefer not to let users configure the "wrong" names in DNS.

Reply:
While I'd prefer the same, there is a cost if we do the conversion in the internal NameResolver. Because the policy name can appear anywhere in the service config, and it can be recursive, we'd need to replace all policy names in the config. There isn't a generic and reliable way to locate them. Because the proto is defined, we could enumerate known locations, but every time a new policy is added to the config proto, we'd need to update the conversion code, in all languages, which could be a hassle.

Alternatively, we could add an internal flag the control whether to accept camel case policy names. We will make it false by default, but turn it on internally. If we decide to go this route, we should probably do the same across all languages.

@markdroth @dfawley @menghanl @AspirinSJL WDTY?

Reply:
Per offline discussion, we have found a way to keep the underscored names in the json converted from proto, and we will not do any normalization in here. "round_robin" will always be "round_robin" everywhere, except for the deprecated LoadBalancingPolicy enum which will still be "ROUND_ROBIN" which is toLowerCase()'ed before used for registry lookup.